### PR TITLE
Fix/upload spreadsheet before name

### LINKED ISSUE #227: Bug: You should be able to upload a spreadsheet before entering the language name
Right now, the app crashes with no indication that anything failed if you uploaded the file before entering a language name.

To fix this, you should be able to upload a file, and secretly, the BCP-47 tag should be set to [`und` for "Undetermined" language](https://iso639-3.sil.org/code/und).

### PR DESCRIPTION
**Pull request by ChrisChrisLoLo**
This fix allows for files to be uploaded before setting a given language. The fix is to set the BCP-47 code as `und` if a file is uploaded before a language is set. 

This fix also does a light refactoring of the Language Select component as well as the Landing Page. The goal of the refactor was to simplify the code down to what is needed, as well as to make the code as "sveltey" (sveltethonic?) as possible.

Changes:

LanguageNameInput.svelte:
- use `bind:` over `createEventDispatcher()`. Hopefully, this is the most idiomatic way of updating the parent component.
- remove the internal `subtext` prop, as there is currently no specific use case for it.
 
LandingPage.svelte:
- refactored logic to make the page work with LanguageNameInput.svelte
- added logic to set the language to "undefined" if a file is uploaded before the language is selected

Fixes #227

~~Will add relevant upload integration tests as soon as #240 gets merged in.~~
EDIT: #240 is now merged in, I will add new integration tests shortly. I have unlisted reviewers in the meantime and converted this issue to a draft. Apologies if I disrupted your workflow!

### CODE REVIEW

**Comment by eddieantonio** [comment]
Hey @ChrisChrisLoLo! I'm confused! Should I review this draft PR? Is it ready for review?

**Comment by ChrisChrisLoLo** [comment]
> Hey @ChrisChrisLoLo! I'm confused! Should I review this draft PR? Is it ready for review?

Hey @eddieantonio! Sorry for the confusion! I listed this as a draft initially to add integration tests today, but if you'd like to review the code now feel free to. I do not anticipate the integration tests to change what I have done. I have made the PR ready for review so as to unblock you or anyone else interested in reviewing the code while I add integration tests.

**Comment by ChrisChrisLoLo** [comment]
Integration tests are now added, and this PR is 100% ready for review!

Apologies for the mixup @eddieantonio! Going forward I'll be sure to _not_ mark anyone as a reviewer until I have added everything I need to. It seems like marking people as a reviewer and then removing them is causing confusion.

File: src/app/components/LanguageNameInput.svelte [hunk 1]
```diff
@@ -103,6 +78,6 @@
     labelFieldName="language" />
   <p class="autocomplete__subtext" data-cy="autocomplete-subtext">
     BCP47Tag:
-    {subtext || ''}
+    {selectedLanguage !== undefined ? selectedLanguage.bcp47Tag : ''}
```
Comment (yehee) [inline src/app/components/LanguageNameInput.svelte]:
```suggestion
    {selectedLanguage?.bcp47Tag ?? ''}
```

File: src/app/pages/LandingPage.svelte [hunk 2]
```diff
@@ -6,21 +6,31 @@
...
+  import type { RelevantKmpOptions } from "@common/kmp-json-file";
 
-  let languageStatus: boolean = false;
+  let selectedLanguage: KeyboardMetadata | undefined = undefined;
   let continueReady: boolean = false;
   let uploadFile: boolean = true;
 
-  $: continueReady = languageStatus && Boolean($currentDownloadURL);
+  $: continueReady =
+    selectedLanguage !== undefined && Boolean($currentDownloadURL);
```
Comment (yehee) [inline src/app/pages/LandingPage.svelte:17]:
```suggestion
    selectedLanguage && Boolean($currentDownloadURL);
```

**Review by yehee — COMMENTED** [review]
I was able to upload a spreadsheet before selecting a language! However, seems like the button to advance to the next page isn't disabled when the language hasn't been selected yet. Also a very minor thing that I noticed that there are a couple of unused imports on `LanguageNameInput.svelte`.

File: src/app/components/LanguageNameInput.svelte (hunk 1 above)
Comment (ChrisChrisLoLo) [reply src/app/components/LanguageNameInput.svelte]: Great suggestion!

File: src/app/pages/LandingPage.svelte (hunk 2 above)
Comment (ChrisChrisLoLo) [reply src/app/pages/LandingPage.svelte:17]: Unfortunately, Typescript complains when I do this due to the comparison with a non-boolean and a boolean. 

I could go `Boolean(selectedLanguage) && Boolean($currentDownloadURL)`, but personally I like being explicit with my types. I feel that casting `selectedLanguage` to a Boolean goes against the explicit nature of typing `selectedLanguage` in the first place.

What are your thoughts?

**Comment by ChrisChrisLoLo** [comment]
> I was able to upload a spreadsheet before selecting a language! However, seems like the button to advance to the next page isn't disabled when the language hasn't been selected yet. Also a very minor thing that I noticed that there are a couple of unused imports on `LanguageNameInput.svelte`.

Thanks for the fast, thorough review! I have removed the useless imports and responded to your feedback. With regards to the button being enabled only when the spreadsheet is uploaded, this is by design. My interpretation of #227 is that the language should be set to `und` before going forward with the next page.

Is there a chance I misinterpreted #227?

**Comment by yehee** [comment]
> Thanks for the fast, thorough review! I have removed the useless imports and responded to your feedback. With regards to the button being enabled only when the spreadsheet is uploaded, this is by design. My interpretation of #227 is that the language should be set to und before going forward with the next page.

Oh no I should have checked #227 before commenting! I just thought that would make sense in terms of overall user experience, but seems like it aligns with what's defined in the issue.

**Comment by eddieantonio** [comment]
>My interpretation of #227 is that the language should be set to und before going forward with the next page.
>
> Is there a chance I misinterpreted #227?

This is the correct interpretation!

File: src/app/components/LanguageNameInput.svelte [hunk 3]
```diff
@@ -103,6 +77,6 @@
     labelFieldName="language" />
   <p class="autocomplete__subtext" data-cy="autocomplete-subtext">
     BCP47Tag:
-    {subtext || ''}
+    {selectedLanguage?.bcp47Tag ?? ''}
```
Comment (eddieantonio) [inline src/app/components/LanguageNameInput.svelte:80]: Nice use of the null-coalescing operator!

File: src/app/pages/LandingPage.svelte [hunk 4]
```diff
@@ -6,21 +6,31 @@
...
   }
 
+  $: if (selectedLanguage !== undefined) {
+    const options: Partial<Readonly<RelevantKmpOptions>> = {
+      languages: [
+        { name: selectedLanguage.language, id: selectedLanguage.bcp47Tag },
+      ],
+    };
+    worker.setProjectData(options);
+  }
```
Comment (eddieantonio) [inline src/app/pages/LandingPage.svelte]: When a reactive statement gets too large, it's best to Extract Function and pass  that function the reactive variable that might change. For example,

```suggestion
  $: updateLanguage(selectedLanguage)

  function updateLanguage(lang: KeyboardMetadata): void {
    if (lang === undefined) return;
    const options: Partial<Readonly<RelevantKmpOptions>> = {
      languages: [
        { name: lang.language, id: lang.bcp47Tag },
      ],
    };
    worker.setProjectData(options);
  }
```

**Review by eddieantonio — APPROVED** [review]
Just one refactoring suggestion, then LGTM!

**EDIT**: the refactors are good! It is, indeed, much more idiomatic!

File: src/app/components/LanguageNameInput.svelte (hunk 3 above)
Comment (ChrisChrisLoLo) [reply src/app/components/LanguageNameInput.svelte:80]: I have to thank @yehee for that one 😅

**Pull request merged**